Move to Python 3

LegitDongo opened the pull request:
## Description
Complete switch to Python 3.7+

Note: Python 2.7 will no longer be supported and anything < Python 3.7 won't work

The version is based on the packages required versions and some of the new python keywords

Also, you will probably want to set up a venv for this

Anybody who previously used Twilio needs to manually update their library:
`pip install twilio==6.45.1`

***Warning*** - copy your cache files before starting PA with this PR! There may be some unexpected edits to them

***This will require extensive testing***

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (would cause existing functionality to change)

## Motivation and Context
Python 2 was discontinued as of January 2020.. We're a little late

## How Has This Been Tested?
I've tested on 3.7 on Windows, tested all types using the webhook tester
No prod test yet

## Wiki Update
- [x] This change requires an update to the Wiki.
- [ ] This change does not require an update to the Wiki.

Wiki changes will come later; we already have a PR about setting up venv; will either add to that or open a new PR for docs


Thanks Sn0opy for the help ❤️

LegitDongo committed:
Move to Python 3

kvangent commented on PokeAlarm/Alarms/__init__.py:
```diff
@@ -27,4 +27,4 @@ def alarm_factory(mgr, settings, max_attempts, api_key):
         return TwitterAlarm(mgr, settings)
     else:
-        raise ValueError("%s is not a valid alarm type!".format(kind))
+        raise ValueError(f"{kind} is not a valid alarm type!")
```
f-strings weren't introduced until python 3.6 - so you need to move your min version up

kvangent commented on PokeAlarm/Load.py:
```diff
@@ -113,10 +113,10 @@ def parse_filter_section(section):
-        local_dts = dict(default_dts.items()
-                         + settings.pop('custom_dts', {}).items())
+    for name, settings in section.pop('filters', {}).items():
+        settings = dict(list(defaults.items()) + list(settings.items()))
```
Have you verified this works right? I'm not sure of the behavior of generator -> `List<Tuple>` -> dict

kvangent replied:
Looks like it might be okay - dict seems to be able to take a list of tuples ok

Might be more efficient/cleaner to use [`itertools.chain`](https://docs.python.org/3/library/itertools.html#itertools.chain) instead of converting though

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -523,7 +523,7 @@ def setup_in_process(self):
 
         # Conect the alarms and send the start up message
-        for alarm in self._alarms.values():
+        for alarm in list(self._alarms.values()):
```
This looks wrong - `dict.values()` should already return a list

kvangent commented on PokeAlarm/Manager.py:
```diff
@@ -746,10 +749,10 @@ def process_stop(self, stop):
         if len(rules) == 0:  # If no rules, default to all
             rules = {"default": Rule(
-                self._stop_filters.keys(), self._alarms.keys())}
+                list(self._stop_filters.keys()), list(self._alarms.keys()))}
```
same with `dict.keys` - should already be a list

kvangent commented on start_pokealarm.py:
```diff
@@ -23,14 +22,12 @@
-reload(sys)
-sys.setdefaultencoding('UTF8')
-
+os.environ["PYTHONIOENCODING"] = 'utf-8'
```
You can remove this line altogether

kvangent commented on tests/filters/test_monster_filter.py:
```diff
@@ -1,16 +1,15 @@
 # ToDo: Find a better way
 # Reinforce UTF-8 as default
-from tests.filters import MockManager, generic_filter_test, full_filter_test
+os.environ["PYTHONIOENCODING"] = 'utf-8'
```
drop this as well

kvangent commented on tools/convert_alarms_file.py:
```diff
@@ -28,48 +28,48 @@ def exchange_set(settings):
     except Exception as e:
-        print ("An unexpected error as occurred: {} - {}"
-               "".format(type(e).__name__, e))
+        print(("An unexpected error as occurred: {} - {}"
```
looks like an extra layer of parenthesis here

kvangent commented on tools/pvp.py:
```diff
@@ -12,15 +12,15 @@ def __init__(self, pa_root):
         multipliers = utils.get_cp_multipliers()
 
-        for json_mon in stats.keys():
+        for json_mon in list(stats.keys()):
```
Shouldn't need list

kvangent commented on tools/pvp.py:
```diff
@@ -12,15 +12,15 @@ def __init__(self, pa_root):
                     = highest["product"]
-                print("{}: highest product at {}: {}".format(
-                    json_mon, limit, highest['product']))
+                print(("{}: highest product at {}: {}".format(
```
extra parenthesis here

kvangent commented on tools/webhook_test.py:
```diff
@@ -8,12 +8,15 @@
 from glob import glob
 
+# Reinforce UTF-8 as default
+os.environ["PYTHONIOENCODING"] = 'utf-8'
```
drop this

kvangent committed:
tests: Disable python2 tests

LegitDongo committed:
Requested changes

LegitDongo committed:
Missed a request

LegitDongo committed:
Remove extra import

LegitDongo committed:
More requested changes

LegitDongo committed:
Bump Twilio lib for 3.7+ support

crhbetz commented:
No issues so far, running 24hrs on production (2 PA instances / 34 MAD devices total). I'll stick with it and report issues asap in case I see anything.

LegitDongo committed:
Update docs to change references from Python 2 to Python 3
Revert gevent to 1.3.7 to prevent C-lib issues
Update dockerfile to use python 3.7
